lib/mix/lib/mix/utils.ex

```diff
@@ -29,17 +29,36 @@ defmodule Mix.Utils do
 
   """
   def mix_config do
-    mix_home_xdg_lookup("XDG_CONFIG_HOME")
+    mix_home_lookup(["MIX_HOME", {:system, :user_config}])
   end
 
-  defp mix_home_xdg_lookup(xdg) do
-    case {System.get_env("MIX_HOME"), System.get_env(xdg)} do
-      {directory, _} when is_binary(directory) -> directory
-      {nil, directory} when is_binary(directory) -> Path.join(directory, "mix")
-      {nil, nil} -> Path.expand("~/.mix")
+  defp mix_home_lookup([]), do: Path.expand("~/.mix")
+
+  defp mix_home_lookup([{:system, type} | rest]) do
+    case xdg?() do
+      true -> xdg_dir(type)
+      _ -> mix_home_lookup(rest)
     end
   end
 
+  defp mix_home_lookup([{env, suffix} | rest]) when is_binary(env) and is_binary(suffix) do
+    case System.get_env(env) do
+      nil -> mix_home_lookup(rest)
+      dir -> Path.join(dir, suffix)
+    end
+  end
+
+  defp mix_home_lookup([env | rest]) when is_binary(env) do
+    mix_home_lookup([{env, ""} | rest])
+  end
+
+  defp xdg_dir(type), do: :filename.basedir(type, "mix", %{os: :linux})
```

Comment:
:rocket: :clap: I'd like to see this support *BSD as well - maybe this could mastch on `{:unix, _} = os.type/0`  ? This is quite common in mix already.

Reply:
Actually it will support BSDs as well. [According to documentation](http://www.erlang.org/doc/man/basedir%252F3) anything other than `'darwin'` and `'windows'` will be interpreted as the same as `'linux'`. So if the naming is confusing I can change that to `:unix`, it will work exactly the same.